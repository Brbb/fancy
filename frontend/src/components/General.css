.langDropdown {
  font-size: 12px;
  height: 35px;
  min-width: 250px;
  border-radius: 10px;
  margin: 10px;
  box-shadow: 1px 1px 0px 2px rgba (0, 0, 0, 0.3);
  background: white;
  border: cornflowerblue solid 1px;
  color: cornflowerblue;
  cursor: pointer;
}

.generalSubmitBtn{
    margin-bottom: 20px
}

.radioGroup {
  width: 250px;
  margin-top: 5px;
  text-align: left;
}

.radioButton{
    font-size: 12px;
    margin-right: 10px
}
