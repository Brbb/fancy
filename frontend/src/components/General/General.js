import React, { Component } from "react";
import {
  Dropdown,
  RadioButtonGroup,
  Button,
  Message
} from "../Elements/Elements";
import { withRouter } from "react-router-dom";
import userApi from "../../services/users/api";

class General extends Component {
  constructor(props) {
    super(props);
    this.state = {
      outcome: { message: "", success: true },
      privacyOptions: [
        { text: "Private", value: "private" },
        { text: "Public", value: "public" }
      ]
    };
  }

  save = async () => {
    let result = await userApi.update(this.props.user);
    if (!result.err) {
      this.setState({ outcome: { message: "Settings saved!", success: true } });
    } else
      this.setState({
        outcome: {
          message: "Error while saving your settings.",
          success: false
        }
      });
  };

  handleLanguageSettingChange = event => {
    var editableUser = this.props.user;
    editableUser.settings.language = event.target.value;
    this.setState({ user: editableUser });
  };

  handlePrivacySettingChange = event => {
    var editableUser = this.props.user;
    editableUser.settings.isPrivate = event.target.value === "private";
    this.setState({ user: editableUser });
  };

  render() {
    return (
      <div className="f-section">
      <label className="f-section-title-no-margin">Preferences</label>
          <hr className="f-hr-section" />
        <Dropdown
          options={this.props.languages}
          text="Language"
          onChange={this.handleLanguageSettingChange}
          selected={this.props.user.settings.language}
        />

        <RadioButtonGroup
          options={this.state.privacyOptions}
          label="Privacy"
          checkedValue={
            this.props.user.settings.isPrivate ? "private" : "public"
          }
          onChange={this.handlePrivacySettingChange}
        />
        <Button
          className="f-submit-button"
          text="Save changes"
          onClick={this.save}
        />
        <Message
          text={this.state.outcome.message}
          success={this.state.outcome.success}
        />
      </div>
    );
  }
}

export default withRouter(General);
