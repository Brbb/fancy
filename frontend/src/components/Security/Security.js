import React, { Component } from "react";
import { Button, InputField, Message } from "../Elements/Basics";
import PasswordGroup from "../Elements/PasswordGroup";
import api from "../../services/users/api";
import { withRouter } from "react-router-dom";

class Security extends Component {
  constructor(props) {
    super(props);
    this.state = {
      oldPassword: "",
      newPassword: "",
      repeatNewPassword: "",
      onUserSettingsChange: () => {},
      error: ""
    };
  }

  handleOldPasswordChange = event => {
    this.setState({ oldPassword: event.target.value });
  };

  save = async () => {
    let result = await api.updatePassword(
      this.props.user,
      this.state.oldPassword,
      this.state.newPassword,
      this.state.repeatNewPassword
    );

    // logout
    if (!result.err) {
      await this.props.onUserSettingsChange();
    } else {
      this.setState({
        error: result.err
      });
    }
  };

  delete = async () => {
    let result = await api.delete(this.props.user);
    if (!result.err) {
      await this.props.onUserSettingsChange();
    } else {
      this.setState({
        error: "User not deleted!"
      });
    }
  };

  render() {
    return (
      <div>
        <div className="f-section">
          <label className="f-section-title-no-margin">Change Password</label>
          <hr className="f-hr-section" />
          <InputField
            name="oldPasswordField"
            type="password"
            placeholder="Old password"
            value={this.state.oldPassword}
            onChange={this.handleOldPasswordChange}
            text="Old Password"
          />
          <PasswordGroup
            onPasswordsChange={(newPassword, repeatPassword, error) => {
              this.setState({
                newPassword: newPassword,
                repeatNewPassword: repeatPassword,
                error: error
              });
            }}
          />
          <Button
            name="save-button"
            className="f-submit-button"
            text="Save Changes"
            onClick={this.save}
            disabled={
              this.state.oldPassword === "" ||
              this.state.newPassword === "" ||
              this.state.repeatNewPassword === "" ||
              this.state.newPassword !== this.state.repeatNewPassword
            }
          />
          <Message text={this.state.error} success={this.state.error === ""} />
        </div>
        <div className="f-section">
          <label className="f-section-title-no-margin warning-text">
            Delete Account
          </label>
          <hr className="f-hr-section" />
          <p className="f-description">
            Once the deletion process begins you will not be able to access to
            your profile and saved settings.
          </p>
          <Button
            name="delete-button"
            className="f-button warning"
            text="Delete"
            onClick={() => {
              if (
                window.confirm(
                  "Are you sure you want to delete your profile?\n(You will be redirected to the login screen)"
                )
              )
                return this.delete();
            }}
          />
        </div>
      </div>
    );
  }
}

export default withRouter(Security);
