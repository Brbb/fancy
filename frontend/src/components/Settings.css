.sidenav {
  height: 100%;
  width: 250px;
  position: fixed;
  z-index: 1;
  top: 0;
  left: 0;
  background-color: aliceblue;
  overflow-x: hidden;
  padding-top: 20px;
}

.sidenavBtn {
  border: cornflowerblue solid 1px;
  color: cornflowerblue;
  height: 40px;
  width: 200px;
  margin:5px
}

.bottomBtn {
  position: absolute;
  left: 30%;
  margin-left: -50px;
  margin-bottom: 30px;
  bottom: 0px;
}
