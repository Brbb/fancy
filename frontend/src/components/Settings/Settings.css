.f-sidenav {
  height: 100%;
  width: 250px;
  position: fixed;
  z-index: 1;
  top: 0;
  left: 0;
  background-color: aliceblue;
  overflow-x: hidden;
  padding-top: 20px;
}

.f-active {
  font-weight: 800;
}

.f-sidenav {
  width: 250px;
  position: fixed;
  z-index: 1;
  top: 0;
  left: 0;
  background: white;
  overflow-x: hidden;
  margin: 20px;
  border-right: solid 1px #e1e4e8;
}

.f-sidenav-button {
  border: #e1e4e8 solid 1px;
  color: #4a4a4a;
  height: 40px;
  width: 200px;
  margin: 5px;
  letter-spacing: 1px;
}

.f-sidenav-logout {
  position: absolute;
  bottom: 0;
  left: 0;
  margin-left: 25px; /*sidenav margin + button margin*/
  margin-bottom: 60px;
}

.f-main {
  margin-left: 50%;
  padding: 0px 10px;
}
