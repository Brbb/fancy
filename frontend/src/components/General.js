import React, { Component } from "react";
import "./General.css";

const languages = [
  "English",
  "French",
  "German",
  "Japanese",
  "Chinese",
  "Korean",
  "Thai"
];

class Security extends Component {
  render() {
    return (
      <div>
        <form className="form" onSubmit={this.handleSubmit}>
          <label className="textFieldLabel">Language</label>
          <select className="langDropdown">
            {languages.map(language => {
              return (
                <option key={language} value={language}>
                  {language}
                </option>
              );
            })}
          </select>
          <label className="textFieldLabel">Privacy</label>
          <div className="radioGroup">
            <label className="radioButton">
              <input type="radio" name="gender" value="public" />
              Public
            </label>
            <label className="radioButton">
              <input type="radio" name="gender" value="private" />
              Private
            </label>
          </div>

          <input
            className="submitBtn generalSubmitBtn"
            type="button"
            value="Save changes"
            onClick={this.login}
          />
        </form>
      </div>
    );
  }
}

export default Security;
