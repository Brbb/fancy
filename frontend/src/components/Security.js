import React, { Component } from "react";
import "./Settings.css";

class Security extends Component {
  
  render() {
    return (
      <div>
      
      </div>
    );
  }
}

export default Security;
