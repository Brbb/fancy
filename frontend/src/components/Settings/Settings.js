import React, { Component } from "react";
import "./Settings.css";
import Security from "../Security/Security";
import General from "../General/General";
import { Button } from "../Elements/Elements";
import authApi from "../../services/auth/api";
import languageApi from "../../services/lang/api";

class Settings extends Component {
  constructor(props) {
    super(props);
    this.state = {
      languages: [],
      error: "",
      currentComponent: "test",
      user: this.props.history.location.state.user
    };
  }

  async componentDidMount() {
    if (!authApi.isAuthorized()) this.props.history.push("/");

    let languages = languageApi.all();
    this.setState({ languages: languages });
  }

  dismissError = () => {
    this.setState({ error: "" });
  };

  handleUserChange = event => {
    this.setState({ username: event.target.value });
  };
  handlePassChange = event => {
    this.setState({ password: event.target.value });
  };

  logout = async () => {
    localStorage.removeItem("jwt");
    await authApi.logout();
    this.props.history.push("/");
  };

  render() {
    let mainAreaComponent =
      this.state.currentComponent === "security" ? (
        <Security user={this.state.user} onUserSettingsChange={this.logout} />
      ) : (
        <General languages={this.state.languages} user={this.state.user} />
      );

    return (
      <div>
        <div className="f-sidenav">
          <Button
            className={
              "f-sidenav-button " +
              (this.state.currentComponent === "general" ? "f-active" : "")
            }
            text="General"
            onClick={() => this.setState({ currentComponent: "general" })}
          />
          <Button
            className={
              "f-sidenav-button " +
              (this.state.currentComponent === "security" ? "f-active" : "")
            }
            text="Security"
            onClick={() => this.setState({ currentComponent: "security" })}
          />
          <Button
            className="f-sidenav-button warning f-sidenav-logout"
            text="Logout"
            onClick={this.logout}
          />
        </div>
        <div className="f-main">{mainAreaComponent}</div>
      </div>
    );
  }
}

export default Settings;
