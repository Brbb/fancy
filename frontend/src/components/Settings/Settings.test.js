import Settings from "./Settings";
import General from "../General/General";
import Security from "../Security/Security";
import authApi from "../../services/auth/api";

describe("Serrings component", () => {
  describe("Rendering tests", () => {
    test("Settings shallow-renders correctly", () => {
      const history = { location: { state: { user: "dummy" } } };
      let wrapper = shallow(<Settings history={history} />);
      expect(wrapper).toMatchSnapshot();
    });

    test("Sidebar buttons switch component correctly between <General /> and <Security />", () => {
      const history = { location: { state: { user: "dummy" } } };

      let wrapper = shallow(<Settings history={history} />);
      expect(wrapper.find(General).length).toEqual(1);
      expect(wrapper.find(Security).length).toEqual(0);

      wrapper
        .find('[name="security-button"]')
        .first()
        .simulate("click");

      expect(wrapper.find(General).length).toEqual(0);
      expect(wrapper.find(Security).length).toEqual(1);

      wrapper
        .find('[name="general-button"]')
        .first()
        .simulate("click");

      expect(wrapper.find(General).length).toEqual(1);
      expect(wrapper.find(Security).length).toEqual(0);
    });

    test("Logout button calls logout function and redirects to <Login />", () => {
      let mockLocation = { state: { user: "dummy" }, pathname: "/settings" };

      let historyMock = {
        location: mockLocation,
        push: jest.fn(path => {
          mockLocation.pathname = path;
        })
      };

      let wrapper = shallow(<Settings history={historyMock} />);

      jest.spyOn(authApi, "logout").mockImplementation(() => {
        return Promise.resolve({});
      });

      expect(authApi.logout).toHaveBeenCalledTimes(0);
      expect(historyMock.location.pathname).toEqual("/settings");

      let logoutButton = wrapper.find('[name="logout-button"]').first();

      return logoutButton
        .prop("onClick")()
        .then(() => {
          expect(authApi.logout).toHaveBeenCalledTimes(1);
          expect(mockLocation.pathname).toEqual("/");
          jest.clearAllMocks();
        });
    });
  });
});
