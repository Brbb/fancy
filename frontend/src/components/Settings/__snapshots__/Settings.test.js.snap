// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Serrings component Rendering tests Settings shallow-renders correctly 1`] = `
<div>
  <div
    className="f-sidenav"
  >
    <Button
      className="f-sidenav-button "
      name="general-button"
      onClick={[Function]}
      text="General"
    />
    <Button
      className="f-sidenav-button "
      name="security-button"
      onClick={[Function]}
      text="Security"
    />
    <Button
      className="f-sidenav-button warning f-sidenav-logout"
      name="logout-button"
      onClick={[Function]}
      text="Logout"
    />
  </div>
  <div
    className="f-main"
  >
    <withRouter(General)
      languages={
        Array [
          "English",
          "French",
          "German",
          "Japanese",
          "Chinese",
          "Korean",
          "Thai",
        ]
      }
      user="dummy"
    />
  </div>
</div>
`;
